addons/mail/models/mail_thread.py

```diff
@@ -1461,6 +1461,21 @@ def message_receive_bounce(self, email, partner, mail_id=None):
         if 'message_bounce' in self._fields:
             for record in self:
                 record.message_bounce = record.message_bounce + 1
+                if record.message_bounce >= 5:
+                    self.env['mail.thread']._message_auto_blacklist(email, partner)
+
+    def _message_auto_blacklist(self, email, partner):
+        recipient_email = '%s <%s>' % (partner.display_name, email)
+        three_months = datetime.datetime.now() - datetime.timedelta(weeks=13)
+        stats = self.env['mail.mail.statistics'].search(
+            [('bounced', '>', fields.Datetime.to_string(three_months))]) \
+            .filtered(lambda x: x.recipient == recipient_email).mapped('bounced')
+        if len(stats) >= 5:
```

Comment:
Why 5 ? 
If there is a true reason, make it a well named constant like `STAT_LIMIT_BEFORE_CRYING = 5`. It will be more usefull for the future.